src/v/cluster/archival_metadata_stm.cc

```diff
@@ -817,4 +874,14 @@ model::offset archival_metadata_stm::get_last_offset() const {
     return _manifest->get_last_offset();
 }
 
+archival_metadata_stm::state_dirty archival_metadata_stm::get_dirty() const {
+    // We are clean if we have written at least one clean record and that
+    // clean record referred to an offset >= the last record that dirtied
+    // the stm.
```

Comment:
Right.  Anything other than a clean command will make it dirty.

This could change in future if we ever added a command type that updated the stm but didn't require manifest updates in the object store, but for now it's more robust to avoid having to explicitly dirty the stm when applying each command.